refactored flush method in producer and added connection error retry and backoff

jmcgrath207 opened the pull request:
Broke up the flush method in the producer to use three different methods ex. `get_batch, _push_kinesis, process_result`

Added retry limit when there is a connection error so it hard fails (usefully for containers) and added a option for exponential_backoff for connection error retries. `get_connection_exponential_backoff`

fixed a test was made to create the kinesis stream. Last change gave me a false positive since the stream was already up in aws. 

Test were ran with `TESTING_USE_AWS_KINESIS=1` with all 34 passing.

Unit test have not been create for this yet, so it's not ready to merge. Just looking for some feedback to make sure I didn't miss something or if I need to break this up into separate pull requests.

hampsterx commented:
neat~ on first glance, looks good! and yep tests and update to readme would be appreciated :)

jmcgrath207 commented:
> @jmcgrath207 this is looking good, can you add const for those stream status. eg
> 
> ```
> RECONNECT = "RECONNECT"
> ...
> ```
> 
> Bit pointless in python I know but still :)

Ya for sure, once I get to point I am happy with it, I will

jmcgrath207 commented:
So this has been pretty difficult to track down but I think I've found my connection issues I am having in us-east-1.  I could only find these when debugging and were hidden inside of boto3. 

One of them is this.

```
2020-11-24 23:42:57,319 DEBUG PID: 9 botocore.parsers Response headers: HTTPHeaderDict({'x-amzn-requestid': 'eb32c5e8-6be5-9b95-b305-59577d4a5982', 'x-amz-id-2': '+bqEpdzsbmNc7wGGSjU42OrM1QVFEqZuKUhb1/HcrpWgPYPmD5XyB/NskBjodMtAo4pQrogGbMTdxlFbNrDyPZDD+UIJrij3PX3NJpXpW80=', 'date': 'Tue, 24 Nov 2020 23:42:57 GMT', 'content-type': 'application/x-amz-json-1.1', 'content-length': '110'})
2020-11-24 23:42:57,320 DEBUG PID: 9 botocore.parsers Response body:
b'{"FailedRecordCount":1,"Records":[{"ErrorCode":"InternalFailure","ErrorMessage":"Internal service failure."}]}'
```
and 

```
b'<ServiceUnavailableException/>\n'
	b'<ServiceUnavailableException/>\n'
2020-11-25 05:48:41,288 DEBUG PID: 8 botocore.hooks Event needs-retry.kinesis.PutRecords: calling handler <botocore.retryhandler.RetryHandler object at 0x7fe07d3e36a0>
2020-11-25 05:48:41,288 DEBUG PID: 8 botocore.retryhandler retry needed: retryable HTTP status code received: 503
2020-11-25 05:48:41,288 DEBUG PID: 8 botocore.retryhandler Reached the maximum number of retry attempts: 1
```

While these are different errors, the behavior I observed is the same. When the error occurs, the Kinesis producer will continue to buffer traffic, however it will stop posting to kinesis all together unless restarted.

So two issue seem to be happening here.
1. It will fail silently due to the `self.flush_task = asyncio.task(self._task)` basically suppressing the raised exception unless the main task checks it. You can see that in the producer put method. 

2. The second one adding the connection manager to rebuild connections that you saw, however while I original thought it was network issue, my issue seems more related to kinesis server side errors. So my theory is I have to recycle the connection when these error are found since the retry wasn't working before. 

Saying that, I am going to be testing this for a week or more before my final commit here but just wanted to keep you in the loop.

And once again, sorry for the large pull request, didn't expect to go down this rabbit hole lol

hampsterx commented:
Interesting. I had found in the past the kinesis service ocassionally seemed to have some odd to understand behaviour that seemed to be from their end rather than the boto library but as I fleshed out this library it seemed to work better and better to the point it seemed like it was under control and I felt confident enough to use it for production services but yeah some of those errors were being suppressed and worse case the container would fall over and restart itself anyway. Not using Kinesis in current role hence haven't worked on this library for a while, was running pretty smooth (ap-southeast) last couple of months I was there though.

Appreciate having someone else take the time to investigate and contribute to the development, Kinesis is a pretty good managed service albeit with a few quirks and makes a suitable alternative to Kafka in some scenarios. As a general rule I think it needs to retry as much as possible but worst case just raise exception and exit. Worse scenario is it hangs/times out as that relies on having monitoring/alerting on the age of the stream to prevent it going unnoticed.

jmcgrath207 commented:
Yeah I think a lot of it has to due to us-east-1 being oversubscribed, in fact shortly after my last post AWS had a 12 hour outage with Kinesis so it might be why I was seeing those error. Fortunately for the async-kinesis project,  I got to add alot of good exception handling for those events, even though I imagine they are pretty rare.

Also I wouldn't feel bad, everything you tested for was good and still believe it's the best async kinesis library out there. I wouldn't want to invest my time otherwise  :nerd_face: 

The only thing I did different in my testing was setup a DNS server(Pihole container) and blocked and unblocked traffic to `kinesis.us-east-1.amazonaws.com`  to test the connection manager code I added (not sure how that could be add to unit testing, but I am open to ideas)

I have also ran these changes in production for a month and they perform well, but I only use the producer in my code.

Everything is working now with all unit tests in python 3.7 and 3.8. Here are the changes.

- Refactored producer code into smaller functions.
- Added connection `retry_limit` argument to Producer and Consumer
- Added connection `expo_backoff` argument to Producer and Consumer
- Added connection `expo_backoff_limit` argument to Producer and Consumer
- Added Connection Manager Logic that is shared between Producer and Consumer see `_get_reconn_helper` and `get_conn` in Base Class
- Invoke Connection Manager Logic to recycle a connection if any connection issue or unknown error issue arise. Basically when it doubt restart the connection. ex https://github.com/jmcgrath207/async-kinesis/blob/master/kinesis/producer.py#L332 https://github.com/jmcgrath207/async-kinesis/blob/master/kinesis/consumer.py#L295

However because of the Connection Manger Logic, It did introduce a breaking change. `create_stream` has been moved to the base class and creation of the stream is done through the instantiation of the object 

```python
Producer(
                stream_name=stream_name, endpoint_url=ENDPOINT_URL,
                create_stream=stream_name, create_stream_shards=2
```

This is because  in `__aenter__()` in the Base class, `self.get_conn()` is called, then `self.start()` which assume the stream exists so you have to tell it at the creation of the object. 

The reason why this was done is so only `self.get_conn` is called for any action that results in a connection state change ex. `ACTIVE, INITIALIZE, RECONNECT` since a lock and a time lock was required to get this working correctly 
https://github.com/jmcgrath207/async-kinesis/blob/master/kinesis/base.py#L159


Let me know what you think and what needs to be changed and thanks in advance for reviewing the worst pull request in history :sweat_smile:

hampsterx commented:
yeah that outage was interesting, gleaned a few details from HackerNews article on it. Code looks sane/great, well done!
Not sure about DNS testing either, I am sure its possible but sounds like overkill, I think the existing test coverage is fine.

Will do major release to reflect breaking change, all good~
